EIPS/eip-signTypedData.md

```diff
@@ -0,0 +1,439 @@
+
+[deterministic]: https://en.wikipedia.org/wiki/Deterministic_algorithm
+[injective]: https://en.wikipedia.org/wiki/Injective_function
+
+### Transactions and bytestrings
+
+An illustrative example of the above breakage can be found in Ethereum. Ethereum has two kinds of messages, transactions `𝕋` and bytestrings `𝔹⁸ⁿ`. These are signed using `eth_sendTransaction` and `eth_sign` respectively. Originally the encoding function `encode : 𝕋 ∪ 𝔹⁸ⁿ → 𝔹⁸ⁿ` was as defined as follows:
+
+* `encode(t : 𝕋) = RLP_encode(t)`
+* `encode(b : 𝔹⁸ⁿ) = b`
+
+While individually they satisfy the required properties, together they do not. If we take `b = RLP_encode(t)` we have a collision. This is mitigated in Geth [PR 2940][geth-pr] by modifying the second leg of the encoding function:
+
+[geth-pr]: https://github.com/ethereum/go-ethereum/pull/2940
+
+* `encode(b : 𝔹⁸ⁿ) = "\x19Ethereum Signed Message:\n" ‖ len(b) ‖ b` where `len(b)` is the ascii-decimal encoding of the number of bytes in `b`.
+
+This solves the collision between the legs since `RLP_encode(t : 𝕋)` never starts with `\x19`. There is still the risk of the new encoding function not being deterministic or injective. It is instructive to consider those in detail.
+
+As is, the definition above is not deterministic. For a 4-byte string `b` both encodings with `len(b) = "4"` and `len(b) = "004"` are valid. This can be solved by further requiring that the decimal encoding of the length has no leading zeros.
+
+The above definition is not obviously collision free. Does a bytestring starting with `"x19Ethereum Signed Message:\n42a…"` mean a 42-byte string starting with `a` or a 4-byte string starting with `2a`?. This was pointed out in [Geth issue #14794][geth-issue-14794] and motivated Trezor to [not implement the standard][trezor] as-is. Fortunately this does not lead to actual collisions as the length of the encoded bytestring provides sufficient information to disambiguate the cases.
+
+[geth-issue-14794]: https://github.com/ethereum/go-ethereum/issues/14794
+[trezor]: https://github.com/trezor/trezor-mcu/issues/163
+
+Both determinism and invectiveness would be trivially true if `len(b)` was left out entirely. The point is, it is difficult to map arbitrary sets to bytestrings without introducing security issues in the encoding function. Yet the current design of `eth_sign` still takes a bytestring as input and expects implementors to come up with an encoding.
```

Comment:
Invectiveness? As in, a tendency to swear? I agree this would be bad in a signing algorithm.